File: PSME/application-ref/src/rest-ref/endpoints/device.cpp
/*!
 * @copyright
 * Copyright (c) 2015-2016 Intel Corporation
 *
 * @copyright
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * @copyright
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * @copyright
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * */

#include "psme/rest-ref/endpoints/device.hpp"
#include "psme/rest-ref/constants/constants.hpp"
#include "agent-framework/module-ref/compute_manager.hpp"
#include <regex>

using namespace psme::rest;
using namespace app::rest::constants;
using namespace app::rest;



namespace {
json::Value make_prototype() {
    json::Value r(json::Value::Type::OBJECT);
    r[Common::ODATA_CONTEXT] = "/redfish/v1/$metadata#Systems/Members/__SYSTEM_ID__/Adapters/Members/__ADAPTER_ID__/Devices/Members/$entity";
    r[Common::ODATA_ID] = json::Value::Type::NIL;
    r[Common::ODATA_TYPE] = "#Device.1.0.0.Device";
    r[Common::ID] = json::Value::Type::NIL;
    r[Common::NAME] = "Device";
    r[Common::DESCRIPTION] = "Device description";
    r[Drive::INTERFACE] = json::Value::Type::NIL;
    r[Drive::CAPACITY] = json::Value::Type::NIL;
    r[Drive::TYPE] = json::Value::Type::NIL;
    r[Drive::BUS_INFO] = json::Value::Type::NIL;
    r[Common::FIRMWARE_VERSION] = json::Value::Type::NIL;
    r[Drive::RPM] = json::Value::Type::NIL;
    r[Common::MANUFACTURER] = json::Value::Type::NIL;
    r[Common::MODEL] = json::Value::Type::NIL;
    r[Common::SERIAL] = json::Value::Type::NIL;
    r[Common::STATUS][Common::STATE] = json::Value::Type::NIL;
    r[Common::STATUS][Common::HEALTH] = json::Value::Type::NIL;
    r[Common::STATUS][Common::HEALTH_ROLLUP] = json::Value::Type::NIL;
    r[Common::OEM] = json::Value::Type::OBJECT;
    r[Common::LINKS][Common::CONTAINED_BY] = json::Value::Type::NIL;
    r[Common::LINKS][Common::OEM] = json::Value::Type::OBJECT;

    return r;
}

}

endpoint::Device::Device(const std::string& path) : Endpoint(path) {}
endpoint::Device::~Device() {}

void endpoint::Device::get(const server::Request& req, server::Response& res) {
    auto r = make_prototype();

    auto system_id = psme::rest::model::Find<agent_framework::model::System>(req.params[PathParam::SYSTEM_ID])
        .get_one()->get_id();
    auto adapter_id = psme::rest::model::Find<agent_framework::model::StorageController>(req.params[PathParam::ADAPTER_ID])
        .via<agent_framework::model::System>(req.params[PathParam::SYSTEM_ID]).get_one()->get_id();

    auto d =  psme::rest::model::Find<agent_framework::model::Drive>(req.params[PathParam::DEVICE_ID])
        .via<agent_framework::model::System>(req.params[PathParam::SYSTEM_ID])
        .via<agent_framework::model::StorageController>(req.params[PathParam::ADAPTER_ID])
        .get();

    r[Common::ODATA_ID] = PathBuilder(req).build();
    r[Common::ODATA_CONTEXT] = std::regex_replace(r[Common::ODATA_CONTEXT].as_string(),
                                                  std::regex("__SYSTEM_ID__"), std::to_string(system_id));
    r[Common::ODATA_CONTEXT] = std::regex_replace(r[Common::ODATA_CONTEXT].as_string(),
                                                  std::regex("__ADAPTER_ID__"), std::to_string(adapter_id));
    r[Common::ID] = req.params[PathParam::DEVICE_ID];

    r[Drive::INTERFACE] = d.get_interface();
    if (d.get_capacity_gb().has_value()) {
        r[Drive::CAPACITY] = endpoint::utils::gb_to_gib(d.get_capacity_gb().value());
    }
    r[Drive::TYPE] = d.get_type();
    r[Drive::BUS_INFO] = d.get_physical_id();
    r[Common::FIRMWARE_VERSION] = d.get_firmware_version();
    r[Drive::RPM] = d.get_rpm();
    r[Common::MANUFACTURER] = d.get_fru_info().get_manufacturer();
    r[Common::MODEL] = d.get_fru_info().get_model_number();
    r[Common::SERIAL] = d.get_fru_info().get_serial_number();
    endpoint::utils::status_to_json(d,r);

    r[Common::LINKS]
     [Common::CONTAINED_BY]
     [Common::ODATA_ID] = PathBuilder(PathParam::BASE_URL)
                            .append(constants::Root::SYSTEMS)
                            .append(req.params[PathParam::SYSTEM_ID])
                            .append(constants::System::STORAGE_ADAPTERS)
                            .append(req.params[PathParam::ADAPTER_ID]).build();

    set_response(res, r);
}

File: PODM/SW/pod-manager/podm-business-entities/src/main/java/com/intel/podm/business/entities/redfish/Device.java
/*
 * Copyright (c) 2016 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.podm.business.entities.redfish;

import com.intel.podm.business.entities.base.DomainObject;
import com.intel.podm.business.entities.base.DomainObjectProperty;
import com.intel.podm.business.entities.redfish.base.Descriptable;
import com.intel.podm.business.entities.redfish.base.Discoverable;
import com.intel.podm.business.entities.redfish.base.LocalStorage;
import com.intel.podm.business.entities.redfish.base.StatusPossessor;
import com.intel.podm.common.types.DriveType;
import com.intel.podm.common.types.Status;
import com.intel.podm.common.types.StorageControllerInterface;

import javax.enterprise.context.Dependent;
import javax.transaction.Transactional;
import java.math.BigDecimal;
import java.net.URI;

import static com.intel.podm.business.entities.base.DomainObjectLink.CONTAINED_BY;
import static com.intel.podm.business.entities.base.DomainObjectLink.OWNED_BY;
import static com.intel.podm.business.entities.base.DomainObjectProperties.decimalProperty;
import static com.intel.podm.business.entities.base.DomainObjectProperties.enumProperty;
import static com.intel.podm.business.entities.base.DomainObjectProperties.integerProperty;
import static com.intel.podm.business.entities.base.DomainObjectProperties.stringProperty;
import static com.intel.podm.common.utils.IterableHelper.single;
import static com.intel.podm.common.utils.IterableHelper.singleOrNull;
import static javax.transaction.Transactional.TxType.REQUIRED;

@Dependent
@Transactional(REQUIRED)
public class Device extends DomainObject implements Discoverable, StatusPossessor, Descriptable, LocalStorage {
    public static final DomainObjectProperty<StorageControllerInterface> INTERFACE = enumProperty("interface", StorageControllerInterface.class);
    public static final DomainObjectProperty<BigDecimal> CAPACITY_GIB = decimalProperty("capacityGib");
    public static final DomainObjectProperty<DriveType> TYPE = enumProperty("type", DriveType.class);
    public static final DomainObjectProperty<Integer> RPM = integerProperty("rpm");
    public static final DomainObjectProperty<String> MANUFACTURER = stringProperty("manufacturer");
    public static final DomainObjectProperty<String> MODEL = stringProperty("model");
    public static final DomainObjectProperty<String> SERIAL_NUMBER = stringProperty("serialNumber");
    public static final DomainObjectProperty<String> FIRMWARE_VERSION = stringProperty("firmwareVersion");
    public static final DomainObjectProperty<String> BUS_INFO = stringProperty("busInfo");

    @Override
    public String getName() {
        return getProperty(NAME);
    }

    @Override
    public void setName(String name) {
        setProperty(NAME, name);
    }

    @Override
    public String getDescription() {
        return getProperty(DESCRIPTION);
    }

    @Override
    public void setDescription(String description) {
        setProperty(DESCRIPTION, description);
    }

    @Override
    public Status getStatus() {
        return getProperty(STATUS);
    }

    @Override
    public void setStatus(Status status) {
        setProperty(STATUS, status);
    }

    public String getBusInfo() {
        return getProperty(BUS_INFO);
    }

    public void setBusInfo(String busInfo) {
        setProperty(BUS_INFO, busInfo);
    }

    public String getFirmwareVersion() {
        return getProperty(FIRMWARE_VERSION);
    }

    public void setFirmwareVersion(String firmwareVersion) {
        setProperty(FIRMWARE_VERSION, firmwareVersion);
    }

    @Override
    public String getSerialNumber() {
        return getProperty(SERIAL_NUMBER);
    }

    public void setSerialNumber(String serialNumber) {
        setProperty(SERIAL_NUMBER, serialNumber);
    }

    public String getModel() {
        return getProperty(MODEL);
    }

    public void setModel(String model) {
        setProperty(MODEL, model);
    }

    public String getManufacturer() {
        return getProperty(MANUFACTURER);
    }

    public void setManufacturer(String manufacturer) {
        setProperty(MANUFACTURER, manufacturer);
    }

    @Override
    public Integer getRpm() {
        return getProperty(RPM);
    }

    public void setRpm(Integer rpm) {
        setProperty(RPM, rpm);
    }

    @Override
    public DriveType getType() {
        return getProperty(TYPE);
    }

    public void setType(DriveType type) {
        setProperty(TYPE, type);
    }

    @Override
    public BigDecimal getCapacityGib() {
        return getProperty(CAPACITY_GIB);
    }

    public void setCapacityGib(BigDecimal capacityGib) {
        setProperty(CAPACITY_GIB, capacityGib);
    }

    @Override
    public StorageControllerInterface getInterface() {
        return getProperty(INTERFACE);
    }

    public void setInterface(StorageControllerInterface anInterface) {
        setProperty(INTERFACE, anInterface);
    }

    public Adapter getAdapter() {
        try {
            return single(getLinked(CONTAINED_BY, Adapter.class));
        } catch (IllegalStateException e) {
            throw new IllegalStateException("Device has to be contained by exactly one Adapter", e);
        }
    }

    @Override
    public URI getSourceUri() {
        return getProperty(SOURCE_URI);
    }

    @Override
    public void setSourceUri(URI sourceUri) {
        setProperty(SOURCE_URI, sourceUri);
    }

    @Override
    public ExternalService getService() {
        return singleOrNull(getLinked(OWNED_BY, ExternalService.class));
    }

    @Override
    public Adapter getParent() {
        return getAdapter();
    }
}
